File: internal/extract/java/testdata/proj/com/foo/Shapes.java
package com.foo;

import java.util.ArrayList;
import java.util.List;

public class Shapes {
    public int constant(int x) {
        return x + 1;
    }

    public int linear(int[] xs) {
        int total = 0;
        for (int x : xs) {
            total += x;
        }
        return total;
    }

    public int quadratic(int[] xs) {
        int n = 0;
        for (int a : xs) {
            for (int b : xs) {
                n++;
            }
        }
        return n;
    }

    public List<Integer> collect(int[] xs) {
        List<Integer> out = new ArrayList<>();
        for (int x : xs) {
            out.add(x * 2);
        }
        return out;
    }

    public int recur(int n) {
        if (n <= 0) {
            return 0;
        }
        return recur(n - 1);
    }

    // noLoopHere: its only loop is inside a lambda body and must not inflate
    // this method's complexity. Regression guard for nested-scope skipping.
    public void noLoopHere(int[] xs) {
        Runnable r = () -> {
            for (int x : xs) {
                System.out.println(x);
            }
        };
        r.run();
    }
}

File: internal/extract/python/testdata/proj/pkg/shapes.py
def constant(x):
    return x + 1

def linear(xs):
    total = 0
    for x in xs:
        total += x
    return total

def quadratic(xs):
    n = 0
    for a in xs:
        for b in xs:
            n += 1
    return n

def collect(xs):
    out = []
    for x in xs:
        out.append(x * 2)
    return out

def recur(n):
    if n <= 0:
        return 0
    return recur(n - 1)

def closure(xs):
    # The loop lives in a nested def and must NOT count toward closure (O(1)).
    def inner():
        s = 0
        for x in xs:
            s += x
        return s
    return inner()
